### ISSUE TITLE
Alternative to ITestOutputHelper, with System.Diagnostics and Essential.Diagnostics

### ISSUE THREAD

**Issue by zijianhuang** [xunit/xunit#412]
I read http://xunit.github.io/docs/capturing-output.html and noticed "it is impossible to know which of the many tests that could be running in parallel were responsible for writing to those shared resources.".

I would think that it is possible to know, if the trace message contains the thread Id or the thread name. 

Obviously it is impractical to demand programmers to include thread Id/Name in the trace messages of the test codes and the library codes. If I could give the thread that launch the test class a name and the name could be shown in front of the trace message, then I should be able to tell which test classes were responsible for writing the messages.

So, basically this is the job of a derived class of System.Diagnostic.TraceListener. And EssentialDiagnostics at http://essentialdiagnostics.codeplex.com (open source, free) had already done the job.

Without using ITestOutputHelper, I could easily have traces output to Windows console with System.Diagnostics.ConsoleTraceListener included in the system.diagnostics section of app.config, and xunit.console.exe can read MyTests.dll.config and ConsoleTraceListener will put trace messages in the test codes and the library codes into Windows console. And I see no good reason why programmers would write Console.Writeline in the test codes, since it is also important to see trace messages, and I consider that the test codes and the library codes should not have knowledge of Windows console.

For tests in parallel, I just need to have these in app.config after having EssentialDiagnostics.dll side by side with the test assembly or in GAC:

```
  <system.diagnostics>
    <switches>
      <add name="SourceSwitch" value="Verbose" />
    </switches>
    <sharedListeners>

      <add name="coloredconsole" type="Essential.Diagnostics.ColoredConsoleTraceListener, Essential.Diagnostics" template="{LocalDateTime:HH':'mm':'ss} [{Thread}] {EventType}: {Message}{Data}" convertWriteToEvent="true" />
      <add name="rollingfile" type="Essential.Diagnostics.RollingFileTraceListener, Essential.Diagnostics" template="{LocalDateTime:HH':'mm':'ss} [{Thread}] {EventType} {Source}: {Message}{Data}" initializeData="%AllUsersProfile%\Fonlow\logs\BasicDataTestsDev_{LocalDateTime:yyyy-MM-dd}_p{ProcessId}.log" />
    </sharedListeners>
    <trace autoflush="true">
      <listeners>
        <add name="rollingfile" />
        <add name="coloredconsole" />
      </listeners>
    </trace>
  </system.diagnostics>
```

The beauty of ColoredConsoleTraceListener is, it renders Trace Information, Warning, Error and Verbose etc. in different colors. And segment "Thread" will output the thread ID by default. In this example, I have both console output, and a trace file for a test run.

If you have many test classes running in parallel, a thread ID might be reused after a test class is finished, and it will be confusing that we identity a test class through a thread ID. It will be good that we could assign the thread a name, ideally the name of the test class when the thread is started.

Currently I may write the following class fixture:

```
    public class ThreadName3Fixture
    {
        public ThreadName3Fixture()
        {
            System.Threading.Thread.CurrentThread.Name = "BasicData";
        }
    }

    public class BasicDataTests : IClassFixture<ThreadName3Fixture>
    {
        [Fact]
```

EssentialDiagnostics will then use the tread name instead of the thread ID.

```
23:03:50 [EntityHelper] Verbose : To create Owner Demo1_16ced404591f4d29a598cbdc309077c2 ...
23:03:50 [BasicEntity] Verbose : To create Owner Demo1_32b264e5d11e468998187715ec3bde91 ...
23:03:50 [BasicData] Verbose : To create Owner Ok_20150504230350766 ...
23:03:57 [BasicData] Verbose : Owner Ok_20150504230350766 initiated.
23:03:57 [BasicEntity] Verbose : Owner Demo1_32b264e5d11e468998187715ec3bde91 initiated.
23:03:57 [BasicEntity] Verbose : To create Owner Demo2_32b264e5d11e468998187715ec3bde91 ...
23:03:57 [EntityHelper] Verbose : Owner Demo1_16ced404591f4d29a598cbdc309077c2 initiated.
23:03:57 [EntityHelper] Verbose : To create Owner Demo2_16ced404591f4d29a598cbdc309077c2 ...
23:03:57 [BasicData] Verbose : To create Owner Ok_20150504230357980 ...
23:03:57 [EntityHelper] Verbose : Owner Demo2_16ced404591f4d29a598cbdc309077c2 initiated.
23:03:57 [BasicEntity] Verbose : Owner Demo2_32b264e5d11e468998187715ec3bde91 initiated.
23:03:58 [BasicData] Verbose : Owner Ok_20150504230357980 initiated.
23:03:58 [BasicEntity] Warning : System.Data.Entity.Infrastructure.DbUpdateException: An error occurred while updating the entries. See the inner exception for details. ---> System.Data.Entity.Core.UpdateException: An error occurred while updating the entries. See the inner exception for details. ---> MySql.Data.MySqlClient.MySqlException: Duplicate entry 'Friend-252' for key 'Tag_Name_idx'

```

To make my test codes shorter and less complex, it will be good that the xUnit team could provide such attribute to decorate a test class:

```
[ThreadName("BasicData")]
public class BasicDataTests

or 

[ThreadName]//the thread name will be BasicDataTests
public class BasicDataTests

```

BTW, I am a coauthor of Essential.Diagnostics.

**Comment by bradwilson** [comment]
I'm sorry, but this really isn't something we're interested in.

The point of addressing why we don't capture Console/Trace/Debug any more is because we used to do those things. Aside from technical difficulties involved with it, we also determined that it was inappropriate to co-opt those production code features for unit testing purposes.

Asking people to use this library instead of `ITetOutputHelper` really violates the latter rule, in addition to simply being a larger, more complex solution than the one we already have.

**Comment by zijianhuang** [comment]
I see that xUnit is primarily for unit testing, and I had well followed how you had made design decisions from v1.9 to v2.0. And I really like the simplicity and elegance of xUnit.

xUnit is also excellent for constructing integration tests.

In library codes, typically you will have:

```
try
{
//something
}
catch (Exception e)
{
    Trace.TraceWarning(e.ToString())
    throw;
}

try
{
//something
}
catch (SpecificException e)
{
    Trace.TraceWarning(e.ToString())
    throw;
}

It is good to see the output of the traces in library codes, as well as the behavior of libraries codes in integration / ordination, not necessarily in production. The limitation of ITestOutputHelper (probably by design, well intended) is that it cannot capture the traces in library codes. I see these trace messages in library codes are not important to unit testing, however they are good for integration testing.

The examples above means programmers do not need to alter existing test codes, just plugin System.Diagnostics which is standard in .NET framework through app.config, and xUnit.Console.exe does not block the output of ConsoleTraceListener. For example, when migrating from v1.9 to v2.0, all I had done was just replacing ITestFixture with IClassFixture, and the behaviors of the test suits in the console runner remain the same.

For tests in parallel, as I had shown, I have already been doing well with a simple class fixture for altering the thread name, and my codes have no knowledge of Essential.Diagnostics. My suggestion may just eliminate the need of writing such class fixture.

It is quite good that xUnit 2.x does not capture Console/Trace/Debug, and I fully agree and support such design, since capturing Console/Trace/Debug is the job of derived classes of TraceListener, which the test codes and the library codes shall have no knowledge.
